batch2/batch/worker.py

```diff
@@ -676,11 +675,32 @@ def __init__(self):
             except Exception as e:
                 if isinstance(e, aiohttp.ClientResponseError) and e.status == 404:   # pylint: disable=no-member
                     raise
-                log.exception(f'failed to mark job ({job_status["batch_id"]}, {job_status["job_id"]}) complete, retrying')
+                log.exception(f'failed to mark {job} complete, retrying')
+
+            # unlist job after 3m or half the run duration
+            now = time.time()
+            elapsed = now - start_time
+            if (job.id in self.jobs and
+                    elapsed > 180.0 and
+                    elapsed > run_duration / 2):
+                log.info(f'too much time elapsed marking {job} complete, removing from jobs, will keep retrying')
+                if job.id in self.jobs:
```

Comment:
You already check above whether job.id is in self.jobs

Reply:
Good catch, fixed.